feat(pallas): add Pallas TPU backend

yarongmu-google opened the pull request:
This commit introduces a new Pallas backend for TPU. This backend allows running PyTorch code on TPU using the Pallas kernel language, without a dependency on `torch_xla`.

The backend is enabled by setting the `PALLAS_TARGET_TPU=1` environment variable. When this variable is set, the Pallas backend will generate code that moves data from CPU tensors to the TPU, executes the Pallas kernel on the TPU, and moves the results back to the CPU.

The implementation includes:
- A new `is_tpu` flag in the Pallas codegen to trigger the TPU-specific logic.
- Data movement using `jax.dlpack.from_dlpack` and `jax.device_get`.
- A new `has_tpu_pallas` utility function to check for the availability of a Pallas-on-TPU environment.
- Validation to ensure that a TPU is available when `PALLAS_TARGET_TPU` is set.
- A new test suite for the Pallas TPU backend, including a test for the validation logic.




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

yarongmu-google committed:
feat(pallas): add Pallas TPU backend

This commit introduces a new Pallas backend for TPU.
This backend allows running PyTorch code on TPU using the Pallas kernel language, without a dependency on `torch_xla`.

The backend is enabled by setting the `PALLAS_TARGET_TPU=1` environment variable.
When this variable is set, the Pallas backend will generate code that moves data from CPU tensors to the TPU, executes the Pallas kernel on the TPU, and moves the results back to the CPU.

The implementation includes:
- A new `is_tpu` flag in the Pallas codegen to trigger the TPU-specific logic.
- Data movement using `jax.dlpack.from_dlpack` and `jax.device_get`.
- A new `has_tpu_pallas` utility function to check for the availability of a Pallas-on-TPU environment.
- Validation to ensure that a TPU is available when `PALLAS_TARGET_TPU` is set.
- A new test suite for the Pallas TPU backend, including a test for the validation logic.

yarongmu-google commented:
@pytorchbot rebase -b main

oulgen commented on torch/_inductor/codegen/pallas.py:
```diff
@@ -527,15 +529,23 @@ def codegen_kernel(self, name: Optional[str] = None) -> str:  # type: ignore[ove
         kernel_name = name or "<KERNEL_NAME>"
         interpret_is_cpu = V.graph.get_current_device_or_throw().type == "cpu"
-        interpret_literal = "True" if interpret_is_cpu else "False"
+        is_tpu = os.environ.get("PALLAS_TARGET_TPU") == "1"
```
shouldnt we get is_tpu from device instead? look at how i did is_gpu/cpu check

oulgen replied:
we do this by device='cpu', and then cpu_backend='pallas'. Device is something user passes in as a model argument and then we can read it via https://github.com/pytorch/pytorch/blob/dc4f3c7505a810322db51e68800b477ed2147947/torch/_inductor/codegen/pallas.py#L427

oulgen commented on requirements-tpu.txt:
```diff
@@ -0,0 +1,15 @@
+
+tpu-info==0.6.0
+
+setuptools==78.1.0
```
?

yarongmu-google replied:
Higher version than in requirements.txt

yarongmu-google committed:
feat(inductor): Enable JIT compilation for Pallas TPU backend

This commit enables JIT compilation (`interpret=False`) for the Pallas TPU backend by changing the tensor data transfer mechanism to work around a bug in the Mosaic compiler.

The previous DLPack-based data transfer method triggered a bug in the Mosaic compiler when handling tensors with 64-bit indexing, which is standard for PyTorch.

This is resolved by changing the data transfer mechanism for the TPU path to first convert the PyTorch tensor to a CPU NumPy array and then use `jax.device_put()` to move it to the TPU. This method avoids the Mosaic bug and allows for successful JIT compilation.

This `device_put` approach is a temporary workaround. Once the TorchTPU project provides a direct and correct way to place a torch.Tensor on a TPU device (similar to the legacy PT/XLA bridge), the `device_put` path should be removed and replaced with the more efficient DLPack method.

With this change, the Pallas TPU backend is now JIT-enabled, and the `interpret=True` workaround has been removed.

yarongmu-google committed:
Removed a few redundant lines.

yarongmu-google commented on test/inductor/test_pallas.py:
```diff
@@ -434,6 +437,35 @@ class PallasTestsCPU(PallasTestsMixin, TestCase):
+
+    def tearDown(self):
+        del os.environ["PALLAS_TARGET_TPU"]
+        super().tearDown()
```
Done.

yarongmu-google committed:
Lint.

yarongmu-google committed:
May revert. This is the xla attempt w/o pt/xla or torch tpu.

oulgen commented on torch/_inductor/codegen/pallas.py:
```diff
@@ -1065,19 +1072,38 @@ def codegen_kernel(self, name: Optional[str] = None) -> str:  # type: ignore[ove
-                    code.writeline(f"{ptr}_jax = jax.dlpack.from_dlpack({ptr})")
+                    if is_tpu:
+                        code.writeline(
+                            f"{ptr}_jax = jax.device_put({ptr}.cpu().numpy(), device=jax.devices('tpu')[0])"
```
should we be hardcoding [0]?

yarongmu-google replied:
For current, initial impl, let's just use [0]. Otherwise we may have much bigger PRs. Let's make sure single chip VM works, then we can expand gradually.

oulgen reviewed (approved):
please address the comments before merging

yarongmu-google committed:
Revert "May revert. This is the xla attempt w/o pt/xla or torch tpu."

This reverts commit 329e3b6dab51621ee67815f27d25864c39d8fe86.

yarongmu-google committed:
Ignore lint for JAX index URL

yarongmu-google committed:
Lint.

voznesenskym commented on torch/_inductor/codegen/pallas.py:
```diff
@@ -1065,19 +1072,38 @@ def codegen_kernel(self, name: Optional[str] = None) -> str:  # type: ignore[ove
-                    code.writeline(f"{ptr}_jax = jax.dlpack.from_dlpack({ptr})")
+                    if is_tpu:
+                        code.writeline(
+                            f"{ptr}_jax = jax.device_put({ptr}.cpu().numpy(), device=jax.devices('tpu')[0])"
```
make this hardcoded and file an issue alongside the PR? something like

TAKE_FIRST_JAX_DEVICE_ONLY = True and raise if it is false with a clear unsupported

jansel commented on requirements-tpu.txt:
This filename is a bit too general, and might confuse people looking for `torch_xla`.  Since this is still experimental, I'd suggest moving this to a setup script with torchinductor in the name.

yarongmu-google committed:
Addressed 2 review comments.

yarongmu-google committed:
Fixed a review comment: rename pallas_target_tpu to _debug_cpu_to_tpu_pallas.

jansel commented:
@pytorchbot merge
